remove user session by user Id

eric-wu opened the pull request:
Please do not merge yet.  It compiles fine and all the tests pass.  But I am still adding more tests especially around CacheProvider.  The core idea is to use Redis transactions to keep track of two mappings, sessionToken --> session and userId --> sessionToken.  Doing so allows us to remove the session by userId.  Please review and let me know your feedback.

Commit:
Add 2 more queued methods to transaction

Commit:
Remove user session by user Id

alxdarksage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -26,41 +34,72 @@ public void setJedisOps(JedisOps jedisOps) {
+            final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+            final String ser = BridgeObjectMapper.get().writeValueAsString(session);
+            results = jedisOps.getTransaction()
+                    .setex(userKey, BridgeConstants.BRIDGE_SESSION_EXPIRE_IN_SECONDS, sessionToken)
```
Users can create more than one session, it looks like your implementation will need to check by user ID on sign in to see if the session already exists, and if so, return that instead of creating a new one... ?

DwayneJengSage replied:
It looks like we already do this:
https://github.com/Sage-Bionetworks/BridgePF/blob/develop/app/controllers/AuthenticationController.java#L29

Eric, can you explicitly test this? Ideally in an integration test, but a manual test will suffice for now.

eric-wu replied:
Indeed the code that does all this is in `AuthenticationController`:

```
    public Result signIn() throws Exception {
        UserSession session = getSessionIfItExists();
        if (session != null) {
            setSessionToken(session.getSessionToken());
            return okResult(new UserSessionInfo(session));
        }
        try {
            JsonNode json = requestToJSON(request());
            SignIn signIn = parseJson(request(), SignIn.class);
            Study study = getStudyOrThrowException(json);
            session = authenticationService.signIn(study, signIn);
            setSessionToken(session.getSessionToken());
            Result result = okResult(new UserSessionInfo(session));
            return result;
        } catch(ConsentRequiredException e) {
            setSessionToken(e.getUserSession().getSessionToken());
            throw e;
        }
    }
```

I believe we're already testing them.

What's new in my change is that session.getUser() is null or user.getId() is null.  They should be exceptions. What I can do is checking on them and throw exceptions.

alxdarksage replied:
This returns the session by GUID, unless you've changed it to look up the session by the user's ID. As a consequence, I can sign in multiple times without re-submittin the GUID I got from my other sessions. Then when deleted/signed out, only one of those sessions will be destroyed (the one whose GUID was submitted for the deletion or signing out). If getSessionIfItExists() is first looking up the user's ID from the GUID, then this should no longer happen, there can only be one session for a given user (user ID).

eric-wu replied:
I see.  Browsing through the code, it looks like the best place to check on this is AuthenticationServiceImpl.getSessionFromAccount() when the user have signed in with Stormpath but haven't created a session yet.

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -26,41 +34,72 @@ public void setJedisOps(JedisOps jedisOps) {
+            final String userKey = RedisKey.USER_SESSION.getRedisKey(userId);
+            final String sessionToken = jedisOps.get(userKey);
+            final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+            jedisOps.getTransaction().del(sessionKey).del(userKey).exec();
```
What if the user is not logged in? I believe this will actually crash with a NullPointerException.

eric-wu replied:
Good catch. I am adding a check on null.

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -129,14 +168,16 @@ public void removeString(String cacheKey) {
-        if (BridgeConfigFactory.getConfig().isLocal()) {
-            throw new BridgeServiceException("Cannot find cache service, have you started a Redis server? (original message: "+e.getMessage()+")");
+    private void promptToStartRedisIfLocal(Throwable e) {
+        if (e.getClass() == JedisException.class && BridgeConfigFactory.getConfig().isLocal()) {
```
Should this be e instanceof JedisException? This way, subclasses of JedisException are caught by this if.

eric-wu replied:
The code being replaced catches Throwable.  I'm limiting it to JedisException.  Still not ideal but should be less noisy than the original code.

eric-wu replied:
Actually it also processes JsonProcessingException. I'm removing the check on JedisException.

DwayneJengSage commented:
Test cases we'll want to test (ideally in an integration test):
- signing in
- signing in when already signed in (returns same session)
- signing out
- signing out when already signed out (shouldn't crash)
- deleting a signed in user (user's session token no longer valid)
- deleting a signed out user (shouldn't crash)

Commit:
More tests around CacheProvider

Commit:
Return a session if it exists

Commit:
Reuse existing session

Commit:
More tests

Commit:
Better name of the test method

eric-wu commented:
Updated:

1) More mocked tests on CacheProvider. It now covers all the methods about sessions.
2) Reuse the existing session if it exists.
3) More smoked tests on the services.

Test cases:

1) Signing in -- This is already covered extensively by AuthenticationServiceImplTest
2) Signing in when already signed in -- Covered by AuthenticationServiceImplTest.signInWhenSignedIn(). I added a check to make sure the same session is reused.
3) Signing out -- Added a test AuthenticationServiceImplTest.testSignOut()
4) Signing out when already signed out -- Added a test AuthenticationServiceImplTest.testSignOutWhenSignedOut()
5) Deleting a signed in user -- Covered extensively by StormPathUserAdminServiceTest
6) Deleting a signed out user -- Added a test StormPathUserAdminServiceTest.testDeleteUserWhenSignedOut()

Future work:

1) [BRIDGE-545 Refactor SessionService](https://sagebionetworks.jira.com/browse/BRIDGE-545)
2) [BRIDGE-546 Lock user on signing in](https://sagebionetworks.jira.com/browse/BRIDGE-546)

Commit:
Merge remote-tracking branch 'team/develop' into develop

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -19,62 +27,115 @@
+                    throw new BridgeServiceException("Session storage error.");
+                }
+            }
+        } catch (JsonProcessingException e) {
```
What about JedisException? Do we need to catch that?

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -19,62 +27,115 @@
-    public UserSession getUserSession(String key) {
+    public UserSession getUserSession(final String sessionToken) {
+        final String sessionKey = RedisKey.SESSION.getRedisKey(sessionToken);
+        final String ser = jedisOps.get(sessionKey);
```
What happens if Jedis throws? Should we leave this inside the try block?

eric-wu replied:
Jedis exceptions will be wrapped into BridgeServiceException by downstream service/controller code.  Behind services and controllers we can just pass whatever exception it is.  Actually does it make sense to remove all the try-catch-rethrow in CacheProvider? What do you think, Alx?

eric-wu replied:
Actually let me restore the try-catch-rethrow blocks.  Don't see we are catching and wrapping exceptions into BridgeServiceException in the downstream code.

DwayneJengSage commented on app/org/sagebionetworks/bridge/cache/CacheProvider.java:
```diff
@@ -19,62 +27,115 @@
-    public void removeSession(String key) {
+    public UserSession getUserSessionByUserId(final String userId) {
+        final String userKey = RedisKey.USER_SESSION.getRedisKey(userId);
+        final String sessionToken = jedisOps.get(userKey);
```
Do we need a try-catch around this?

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/StormPathUserAdminServiceTest.java:
```diff
@@ -131,6 +131,13 @@ public void cannotCreateUserWithSameUsernameOrEmail() {
+        UserSession session = userAdminService.createUser(signUp, study, true, true);
+        authService.signOut(session);
+        // Shouldn't crash
+        userAdminService.deleteUser(study, session.getUser().getEmail());
```
Can we validate the user session (or lack thereof) to make sure deleting the user deletes the session?

Commit:
Restore the try-catch blocks

Commit:
Restore the try-catch blocks

Commit:
Add test to verify the session is removed after signing out

DwayneJengSage commented:
Sorry, missed one more. We'll need the same session check in StormPathUserAdminServiceTest.deletedUserHasBeenDeleted()

Commit:
Delete a deleted user should not crash
